Management Command to Clone Web Users

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[JIRA Ticket](https://dimagi-dev.atlassian.net/browse/SS-205)

After determining that a script to update the username "in-place" was too risky given our assumption that a username is constant, this is our next best option. The idea is to make it as easy as possible for an organization to migrate users from an outdated email domain to a new email domain. This is specifically relevant in the context of SSO as a user on an outdated email domain cannot take advantage of it, but this also applies generally as we do not support updating usernames.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
This is "just" a management command, meaning the code is safe to merge, but running the command is a different story.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added some smaller tests around pieces of the management command.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-3681

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
first pass on management command to clone web users

the expected input is a csv with an old_username and new_username column

gherceg committed:
fix failing tests by setting forcing views to be reindexed

gherceg committed:
cleanup (how did I miss this)

gherceg committed:
do not set stale to False in management command

millerdev commented:
> the code is safe to merge, but running the command is a different story.

LOL! Could that be rephrased as _You can merge this, but good luck if you decide to run it_?

> running a manual task to delete the old users

Is there any way to disable the old account? That seems safer than deleting it. If anyone needed to get back in we could temporarily re-enable it, but have no extra maintenance steps if that was never needed.

These are my initial reactions after scanning the discussion. I didn't get time to review the code yet.

biyeun commented:
I'm with @millerdev on disabling accounts vs deleting. I can't imagine that it takes up so much space that we should be concerned about keeping that data around.

biyeun commented on corehq/apps/users/management/commands/clone_web_users.py:
```diff
@@ -0,0 +1,183 @@
+
+def send_deprecation_email(old_user, new_user):
+    context = {
+        'greeting': _("Dear %s,") % new_user.first_name if new_user.first_name else _("Hello,"),
```
nit `_("Dear {first_name},").format(first_name=new_user.first_name)` instead of `%` is preferred

gherceg commented:
I can get behind disabling instead of deleting. To come to a consensus on what that means for this command, I see two main options:
- The Automated Way
  - run the command, create an action item in a sql table to disable the old user that is checked the next time the new user logs in, re-enable the old user temporarily if needed (NOTE: this still requires manual intervention _if_ a user needs to access their old account)
- The Manual Way
  - run the command, run a separate command a week later to disable the list of old users
  
As Joel mentioned, without concrete examples of what a user might need to access their old account for, option 1 does seem worth the extra effort/sql table. So at the moment I am leaning towards the automated way, but if anyone has concerns please voice them.

gherceg committed:
Merge branch 'master' into gh/clone-users

millerdev commented:
I agree that option 1 sounds like the best of those two options. But it seems like it would be even better if we did not need to implement the check to disable the old account since I assume that check would need to be done for every user on login, not only ones that have been cloned.

Would it be possible to get a workflow approved where the accounts are cloned and the old ones are immediately disabled?

gherceg commented:
> that check would need to be done for every user on login, not only ones that have been cloned

That is correct.

As far as disabling users immediately, I think from a logistics standpoint running this one weekend and ensuring MSI is aware of when this will happen is completely valid. 

My concern is that this script will have missed migrating over something, and support will be dealing with requests to access their old user. Though I suppose this hypothetical scenario is possible whether or not we disable users immediately, so it is better to avoid the additional overhead of a new table.

gherceg committed:
set stale to false for saved reports and scheduled reports queries

this is solely to ensure tests pass, but since this is a management
command that is not run often, it should be fine to always have stale
set to false

gherceg commented on corehq/apps/users/tests/test_clone_web_users.py:
```diff
@@ -0,0 +1,127 @@
+        actual_id = ReportNotification.by_domain_and_owner(self.domain, self.new_user._id, stale=False)[0]._id
+        self.assertEqual(expected_id, actual_id)
+
+    def test_transfer_saved_reports(self):
+        saved_report = ReportConfig(
+            name='test',
+            owner_id=self.old_user._id,
+            report_slug='worker_activity',
+            report_type='project_report',
+            domain=self.domain)
+        saved_report.save()
+        expected_id = saved_report._id
+        self.addCleanup(saved_report.delete)
+        transfer_saved_reports(self.old_user, self.new_user)
+
+        actual_id = ReportConfig.by_domain_and_owner(self.domain, self.new_user._id, stale=False)[0]._id
+        self.assertEqual(expected_id, actual_id)
```
Thanks for mentioning the flaky test possibility. It should have crossed my mind when thinking about that solution, and agree it would make the test less readable. I set stale to False again in the saved report and scheduled report queries. I think about adding it as an optional param defaulted to True that tests could override, but given that this is a management command that isn't run often, I think it is best to have stale set to false anyway.

gherceg commented on corehq/apps/users/management/commands/clone_web_users.py:
```diff
@@ -0,0 +1,183 @@
+
+    def iterate_usernames_to_update(self, file):
+        with open(file, newline='') as csvfile:
+            reader = csv.reader(csvfile, delimiter=' ', quotechar='|')
+            header = next(reader)
+            header_values = header[0].split(',')
+            try:
+                old_user_index = header_values.index(OLD_USERNAME)
+                new_user_index = header_values.index(NEW_USERNAME)
+            except ValueError:
+                logger.error(f'The provided csv file should contain both a {OLD_USERNAME} and {NEW_USERNAME} '
+                             'header column')
+                return
+            for row in reader:
+                row_values = row[0].split(',')
+                yield row_values[old_user_index], row_values[new_user_index]
```
😍

gherceg committed:
use csv DictReader

gherceg committed:
use .format instead of % for string

gherceg committed:
set is_active to False for old user

gherceg committed:
transfer feature flags

gherceg committed:
update is_active on django user

gherceg committed:
set location_id directly

gherceg committed:
fix email template issue and update text

gherceg committed:
only save new user after creation is successful

gherceg committed:
set is_admin when copying domain memberships

gherceg committed:
better logging

gherceg commented on corehq/apps/users/management/commands/clone_web_users.py:
```diff
@@ -75,6 +75,12 @@ def create_new_user_from_old_user(old_user, new_username):
 
 
+def deactivate_user(user):
+    user.is_active = False
```
Thanks for raising this. It looks like we include the ModelBackend in AUTHENTICATION_BACKENDS which the docs mention enforce the `is_active` flag. Also while locally testing, I confirmed that my "old user" was logged out after running this command because of this flag. It is interesting to think that if in the future we swap out django authentication backends for whatever reason, inactive users would have to be dealt with properly. It seems safe enough to assume we will handle that situation appropriately if/when we get there.

gherceg committed:
fix broken test

biyeun reviewed (approved):
changes look great! 🎉

gherceg committed:
add error handling for attempting to create a user that already exists

gherceg committed:
use get_email()

gherceg committed:
add better docs for command

gherceg committed:
update email text

joxl commented on corehq/apps/users/management/commands/clone_web_users.py:
```diff
@@ -0,0 +1,215 @@
+
+
+def iterate_usernames_to_update(file):
+    with open(file, newline='') as csvfile:
```
nit: (I must have missed this on my first pass, but not a blocker). I don't think `csv` readers care what flavor of universal newlines the file is opened in, so the `newline=''` can be omitted. Setting the newline mode can lead to confusion since it appears like a non-standard CSV format might be expected, which I don't think is the case here.

gherceg committed:
remove newline arg when opening csv file

gherceg committed:
collect already existing users and print at end

gherceg committed:
validate new username is an email

gherceg committed:
handle non existent old usernames; improve logging

gherceg committed:
Merge branch 'master' into gh/clone-users

gherceg committed:
lowercase old_username before looking up

gherceg committed:
copy email from old user instead of using new username

gherceg committed:
fix bug in transfer_feature_flags

- was using toggle name in set_toggle when it should be toggle slug

gherceg committed:
add support for dry-run arg

gherceg committed:
add NewUserAlreadyExists exception to catch during dry runs

gherceg committed:
add structure for notify command

gherceg committed:
pull out method for validating usernames

gherceg committed:
implement notify command

gherceg committed:
consolidate iteration and collecting metrics logic

gherceg committed:
update/simplify help docs

gherceg committed:
add templates for scheduled migration notification email

gherceg committed:
fix dry-run bug

gherceg commented:
Context on all of the commits I just pushed. We decided a few things:
- users should be given some notice this is happening (shoutout Biyeun), and since the logic is similar to send out emails to impacted users, I created specific commands to handle both giving the users a heads up (`notify`), and actually running the migration (`run`)
- I don't know why I didn't add a dry-run flag sooner, but it's here now! 

That's essentially what the recent round of commits entail.

gherceg committed:
add dry run tag to logs

gherceg committed:
fix test issue with transfer_feature_flags

I swear I ran this test locally but obviously not

gherceg committed:
pass entire new_user object to transfer_scheduled_reports to handle dry-run case

gherceg committed:
fix transfer_feature_flag issue

gherceg committed:
revert transfer_feature_flag changes

gherceg committed:
format date

gherceg committed:
update email content

gherceg committed:
use <ul> tag for html emails

gherceg committed:
ensure mobile users are not supported

millerdev commented on corehq/apps/users/templates/users/email/deprecated_user.txt:
```diff
@@ -0,0 +1,24 @@
+ - Pending CommCare invitations
+
+The {{ old_username }} user is no longer active. You can log into your new user with the username
+{{ new_username }}. If you are exempt from SSO, you can follow the Forgot Password workflow to successfully login.
```
Is the meaning of "SSO" commonly known? Would it be worth spelling it out rather than using an acronym?

The same also applies to other templates.

biyeun replied:
reminder email text updated after feedback from Erin. This line was scrapped as it's highly unlikely this will apply to any user needing a migration. The users that fall into that category are account admins

biyeun committed:
update text and formatting for reminder email

millerdev commented on corehq/apps/users/templates/users/email/notify_scheduled_user_migration.html:
```diff
@@ -4,31 +4,30 @@
-    organization has requested that your user, {{ old_username }}, be migrated to a new user with the username
-    {{ new_username }}.
-
+    We have scheduled a migration to be run on {{ scheduled_date }} that will impact your CommCare user. MSI has requested that your user, {{ old_username }}, be migrated to a new user with the username {{ new_username }}.
```
Are we planning to remove these templates after using them for MSI? It doesn't seem right to encode a specific customer name here.

The same applies to other templates as well.

gherceg committed:
Merge branch 'master' into gh/clone-users

gherceg commented:
This work has been complete, and we have no plans to merge this into master. I edited the commit to remove the reference to a specific organization in an email template which is why I force pushed. I will close this PR after tests complete as we don't need it in HQ, but it might come in handy again some day.
